lib/pharos/phases/configure_dns.rb

```diff
@@ -45,52 +46,53 @@ def max_unavailable
 
       # @param replicas [Integer]
       # @param nodes [Integer]
-      def patch_kubedns(replicas:, max_surge:, max_unavailable:)
-        logger.info { "Patching kube-dns addon with #{replicas} replicas (max-surge #{max_surge}, max-unavailable #{max_unavailable})..." }
+      def patch_deployment(name, replicas:, max_surge:, max_unavailable:)
+        logger.info { "Patching #{name} deployment with #{replicas} replicas (max-surge #{max_surge}, max-unavailable #{max_unavailable})..." }
 
-        resource = Pharos::Kube.session(@master.api_address).resource(
-          apiVersion: 'extensions/v1beta1',
-          kind: 'Deployment',
-          metadata: {
-            namespace: 'kube-system',
-            name: 'kube-dns'
-          },
-          spec: {
-            replicas: replicas,
-            strategy: {
-              type: "RollingUpdate",
-              rollingUpdate: {
-                maxSurge: max_surge, # must be zero for a two-node cluster
-                maxUnavailable: max_unavailable, # must be at least one, even for a single-node cluster
-              }
-            },
-            template: {
-              spec: {
-                affinity: {
-                  podAntiAffinity: {
-                    requiredDuringSchedulingIgnoredDuringExecution: [
-                      {
-                        labelSelector: {
-                          matchExpressions: [
-                            {
-                              key: "k8s-app",
-                              operator: "In",
-                              values: [
-                                "kube-dns"
-                              ]
-                            }
-                          ]
-                        },
-                        topologyKey: "kubernetes.io/hostname"
-                      }
-                    ]
-                  }
+        Pharos::Kube.session(@master.api_address).resource_client('apps/v1').patch_deployment(
```

Comment:
Patch should be ok here, right?

Reply:
Yes, the `PATCH` matches what `kubectl set image` does:

```
$ kubectl -v8 -n kube-system set image deployments/coredns coredns=quay.io/kontena/coredns-amd64:1.1.3
...
I0807 11:11:05.700941    8436 request.go:874] Request Body: {"spec":{"template":{"spec":{"$setElementOrder/containers":[{"name":"coredns"}],"containers":[{"image":"quay.io/kontena/coredns-amd64:1.1.3","name":"coredns"}]}}}}
I0807 11:11:05.700985    8436 round_trippers.go:383] PATCH https://167.99.39.233:6443/apis/extensions/v1beta1/namespaces/kube-system/deployments/coredns
...
```

Not entirely sure what the `$setElementOrder` is doing, but it seems unnecessary. Apart from that the container image part is identical.

Reply:
Just a note that this also changes how dns replicas/maxSurge/etc is sent.

Reply:
Yes, those are simple object-level PATCHes, no array merge needed. Verified that the end result is as intended.